provider/pkg/gen/java-templates/src/main/java/com/pulumi/kubernetes/apiextensions/CustomResource.java

```diff
@@ -0,0 +1,220 @@
+     */
+    public CustomResource(String name) {
+        this(name, CustomResourceArgs.Empty);
+    }
+
+    /**
+     *
+     * @param name The _unique_ name of the resulting resource.
+     * @param args The arguments to use to populate this resource's properties.
+     */
+    public CustomResource(String name, @Nullable CustomResourceArgsBase args) {
+        this(name, args, null);
+    }
+    
+    /**
+     *
+     * @param name The _unique_ name of the resulting resource.
+     * @param args The arguments to use to populate this resource's properties.
+     * @param options A bag of options that control this resource's behavior.
+     */
+    public CustomResource(String name, @Nullable CustomResourceArgsBase args, @Nullable com.pulumi.resources.CustomResourceOptions options) {
+        super(makeType(args), name, makeArgs(args), makeResourceOptions(options, Codegen.empty()));
+    }
+
+    // private CustomResource(String name, Output<String> id, @Nullable com.pulumi.resources.CustomResourceOptions options) {
+    //     super(makeType(args), name, null, makeResourceOptions(options, id));
+    // }
+
+    private static String makeType(@Nullable CustomResourceArgsBase args) {
+        String apiVersion = args.apiVersion().map(Internal::of).map(CustomResource::getOutputValue).orElse("");
+        String kind = args.kind().map(Internal::of).map(CustomResource::getOutputValue).orElse("");
+        return String.format("kubernetes:%s:%s", apiVersion, kind);
+    }
+
+    private static String getOutputValue(OutputInternal<String> o) {
+        try {
+            return o.getValueOrDefault("").get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static class Field {
+        public Optional<?> value;
+        public Either<Annotation,AnnotationDescription> importAnnotation;
+
+        public Field(Optional<?> value, Either<Annotation,AnnotationDescription> importAnnotation) {
+            this.value = value;
+            this.importAnnotation = importAnnotation;
+        }
+    }
+
+    private static ResourceArgs makeArgs(@Nullable CustomResourceArgsBase args) {
+        if (args == null) {
+            return null;
+        }
+
+        // collect the input properties from the annotated fields of the user-supplied args,
+        // plus the dynamic input properties within otherFields.
+        var imports = ImportMetadata.of(args.getClass());
+        var importFields = 
+            imports.values().stream()
+            .map(info -> new Field(info.getFieldOutput(args), Either.ofLeft(info.getAnnotation())));
+        var otherFields = 
+            args.otherFields().orElseGet(Map::of).entrySet().stream()
+            .map(entry -> new Field(Optional.ofNullable(entry.getValue()), Either.ofRight(
+                AnnotationDescription.Builder.ofType(Import.class).define("name", entry.getKey()).build())));
```

Comment:
This code annotates the generated field with an `@Import(name="foo")` annotation.  Should the `json` flag be set on the `@Import` annotation?

Reply:
I've been diving back into the java codebase to figure this out, this is really not annotated properly.. (and why is it called Import? Sounds like InputProperty would be a better name).

The json=true is set for explicit providers args only. For resources it's not set. 

https://github.com/pulumi/pulumi-java/blob/main/pkg/codegen/java/gen.go#L449

I believe there is (was?) a problem in Pulumi where provider Configure/CheckConfig methods did not properly accept a normal encoding of values, instead using a json-in-proto-encoding coming from the SDK, and Java emulates that.

Reply:
For custom resources, do not use it.

Reply:
Thanks @t0yv0, that tracks.